[Cocoa] Genmoji viewed in Gmail and other web mail clients are very large

pxlcoder opened the pull request:
#### 3102843f9dd6361edda9a9e824fc17bca82b0ad4
<pre>
[Cocoa] Genmoji viewed in Gmail and other web mail clients are very large
<a href="https://bugs.webkit.org/show_bug.cgi?id=283202">https://bugs.webkit.org/show_bug.cgi?id=283202</a>
<a href="https://rdar.apple.com/137165281">rdar://137165281</a>

Reviewed by Abrar Rahman Protyasha.

In places where Genmoji aren&apos;t fully supported, a fallback PNG is displayed
instead. Currently, this fallback is created using the first frame of the
HEIC file that represents a Genmoji. This is problematic, since the size of
the first frame can be inconsistent, and can be quite large, looking very
out of place alongside text.

To fix, create the PNG at a much smaller fixed size. Note that this can
still result in inconsistencies between font size / fallback size, if the
text itself is larger. That issue is out of scope for this change, since
the only general solution is to recreate a PNG every time the used font size
is changed. A partial solution is to offer a post-processing step prior to
sending out an email, but that would only benefit clients that knew to
perform the post-processing step (potentially Apple Mail).

* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::fallbackImageForMultiRepresentationHEIC):
(WebCore::Editor::insertMultiRepresentationHEIC):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/AdaptiveImageGlyph.mm:
(TestWebKitAPI::TEST(AdaptiveImageGlyph, InsertAndRemoveWKAttachments)):
(TestWebKitAPI::TEST(AdaptiveImageGlyph, InsertWKAttachmentsCopyFromWebViewPasteToWebView)):
(TestWebKitAPI::TEST(AdaptiveImageGlyph, InsertWKAttachmentsMovingParagraphs)):

Canonical link: <a href="https://commits.webkit.org/286692@main">https://commits.webkit.org/286692@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/106ae5ab6722d5bc319abe83b68a349125339f9c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76726 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55761 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29632 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81258 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28002 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78843 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64903 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4054 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60125 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18221 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79793 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50065 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65866 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40447 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47467 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23360 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26326 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68585 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23690 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82703 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4102 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2707 "Found 1 new test failure: imported/w3c/web-platform-tests/encoding/encodeInto.any.serviceworker.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68409 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4255 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65838 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67661 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11634 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9712 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11882 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4049 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6858 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4072 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7502 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5830 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/editing/cocoa/EditorCocoa.mm:
```diff
@@ -394,6 +394,43 @@ static void maybeCopyNodeAttributesToFragment(const Node& node, DocumentFragment
 
 #if ENABLE(MULTI_REPRESENTATION_HEIC)
+
+static RetainPtr<CGImageRef> fallbackImageForMultiRepresentationHEIC(const std::span<const uint8_t>& data)
```
`std::span` is a lightweight object so you can pass it by value, by the way.

aprotyas replied:
i.e. 
```suggestion
static RetainPtr<CGImageRef> fallbackImageForMultiRepresentationHEIC(std::span<const uint8_t> data)
```

pxlcoder replied:
Done!

pxlcoder committed:
[Cocoa] Genmoji viewed in Gmail and other web mail clients are very large
https://bugs.webkit.org/show_bug.cgi?id=283202
rdar://137165281

Reviewed by Abrar Rahman Protyasha.

In places where Genmoji aren't fully supported, a fallback PNG is displayed
instead. Currently, this fallback is created using the first frame of the
HEIC file that represents a Genmoji. This is problematic, since the size of
the first frame can be inconsistent, and can be quite large, looking very
out of place alongside text.

To fix, create the PNG at a much smaller fixed size. Note that this can
still result in inconsistencies between font size / fallback size, if the
text itself is larger. That issue is out of scope for this change, since
the only general solution is to recreate a PNG every time the used font size
is changed. A partial solution is to offer a post-processing step prior to
sending out an email, but that would only benefit clients that knew to
perform the post-processing step (potentially Apple Mail).

* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::fallbackImageForMultiRepresentationHEIC):
(WebCore::Editor::insertMultiRepresentationHEIC):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/AdaptiveImageGlyph.mm:
(TestWebKitAPI::TEST(AdaptiveImageGlyph, InsertAndRemoveWKAttachments)):
(TestWebKitAPI::TEST(AdaptiveImageGlyph, InsertWKAttachmentsCopyFromWebViewPasteToWebView)):
(TestWebKitAPI::TEST(AdaptiveImageGlyph, InsertWKAttachmentsMovingParagraphs)):

Canonical link: https://commits.webkit.org/286692@main
